### PR TITLE
Reshape the Dockerfile for clickhouse/clickhouse-server

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Split huge `RUN` in Dockerfile into smaller conditional. Install the necessary tools on demand in the same `RUN` layer, and remove them after that. Upgrade the OS only once at the beginning. Use a modern way to check the signed repository. Downgrade the base repo to ubuntu:20.04 to address the issues on older docker versions. Upgrade golang version to address golang vulnerabilities

### Links to issues with docker image
https://github.com/ClickHouse/ClickHouse/issues/48296
https://github.com/ClickHouse/ClickHouse/issues/50740
https://github.com/ClickHouse/ClickHouse/issues/49883
https://github.com/ClickHouse/ClickHouse/pull/50958
https://github.com/ClickHouse/ClickHouse/issues/51247

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 190be181190039101c30a6991fecca945250958a with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51504/190be181190039101c30a6991fecca945250958a/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by Felixoid** [comment]
I am waiting for the release build to check the `docker scout` with upgraded golang.

Now when I build the images as following:

```
git checkout shape-up-docker-server
DOCKER_BUILDKIT=1 docker build -t clickhouse/clickhouse-server:head-20.04 --file=./$(git rev-parse --show-cdup)/docker/server/Dockerfile.ubuntu ./$(git rev-parse --show-cdup)/docker/server
git checkout master
DOCKER_BUILDKIT=1 docker build -t clickhouse/clickhouse-server:head-22.04 --file=./$(git rev-parse --show-cdup)/docker/server/Dockerfile.ubuntu ./$(git rev-parse --show-cdup)/docker/server
```

I have the next results for the commands:

```
docker scout cves --output=/tmp/head-22.04.txt clickhouse/clickhouse-server:head-22.04
docker scout cves --output=/tmp/head-20.04.txt clickhouse/clickhouse-server:head-20.04
```

[head-20.04.txt](https://github.com/ClickHouse/ClickHouse/files/11887620/head-20.04.txt)
[head-22.04.txt](https://github.com/ClickHouse/ClickHouse/files/11887621/head-22.04.txt)

I don't see a reason for keeping the broken 22.04 if we can fix it and have a working image.

cc @juliojimenez

**Review by alexey-milovidov — APPROVED** [review]

File: docker/server/Dockerfile.ubuntu [hunk 1]
```diff
@@ -11,17 +11,17 @@ RUN sed -i "s|http://archive.ubuntu.com|${apt_archive}|g" /etc/apt/sources.list
     && apt-get update \
     && apt-get upgrade -yq \
     && apt-get install --yes --no-install-recommends \
-        apt-transport-https \
-        ca-certificates \
```
Comment (Felixoid) [inline docker/server/Dockerfile.ubuntu:15]: oh no...

**Comment by Felixoid** [comment]
Despite the fact the docker image build fails, I can continue the experiment. Here it is:

```
git checkout master
docker build -t clickhouse/clickhouse-server:head-22.04 --build-arg=REPOSITORY='https://s3.amazonaws.com/clickhouse-builds/PRs/51504/190be181190039101c30a6991fecca945250958a/package_release' --build-arg=deb_location_url='https://s3.amazonaws.com/clickhouse-builds/PRs/51504/190be181190039101c30a6991fecca945250958a/package_release' --build-arg=VERSION='23.6.1.1325' --file=./$(git rev-parse --show-cdup)/docker/server/Dockerfile.ubuntu ./$(git rev-parse --show-cdup)/docker/server
docker scout cves --output=/tmp/head-22.04.txt clickhouse/clickhouse-server:head-22.04
    ✓ Image stored for indexing
    ✓ Indexed 203 packages
    ✗ Detected 8 vulnerable packages with a total of 8 vulnerabilities
    ✓ Report written to /tmp/head-22.04.txt
git checkout shape-up-docker-server
docker build -t clickhouse/clickhouse-server:head-20.04 --build-arg=REPOSITORY='https://s3.amazonaws.com/clickhouse-builds/PRs/51504/190be181190039101c30a6991fecca945250958a/package_release' --build-arg=deb_location_url='https://s3.amazonaws.com/clickhouse-builds/PRs/51504/190be181190039101c30a6991fecca945250958a/package_release' --build-arg=VERSION='23.6.1.1325' --file=./$(git rev-parse --show-cdup)/docker/server/Dockerfile.ubuntu ./$(git rev-parse --show-cdup)/docker/server
docker scout cves --output=/tmp/head-20.04.txt clickhouse/clickhouse-server:head-20.04
    ✓ Image stored for indexing
    ✓ Indexed 188 packages
    ✗ Detected 7 vulnerable packages with a total of 8 vulnerabilities
    ✓ Report written to /tmp/head-20.04.txt
```

[head-20.04.txt](https://github.com/ClickHouse/ClickHouse/files/11892060/head-20.04.txt)
[head-22.04.txt](https://github.com/ClickHouse/ClickHouse/files/11892061/head-22.04.txt)

It shows that using 22.04 is not requirement from the security perspective, but we'll address a huge compatibility issue.

**Comment by Felixoid** [comment]
```
docker scout cves --output=/tmp/head-20.04.txt clickhouse/clickhouse-server:head-20.04
    ✓ Image stored for indexing
    ✓ Indexed 188 packages
    ✗ Detected 7 vulnerable packages with a total of 8 vulnerabilities
    ✓ Report written to /tmp/head-20.04.txt
```

The image looks good from my POV. I am merging it to have fixed images in the coming and previous releases.

**Pull request merged**